File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        combination = []
        res = []

        if candidates[0] > target:
            return res

        def backtracking(idx, currSum):
            if currSum == target:
                res.append(copy.copy(combination))
                return

            for i in range(idx, len(candidates)):
                if currSum + candidates[i] <= target:
                    combination.append(candidates[i])
                    backtracking(i, currSum + candidates[i])
                    combination.pop()
                else:
                    break

        backtracking(0, 0)
        return res

File: 0039-combination-sum/0039-combination-sum.ts
function combinationSum(candidates: number[], target: number): number[][] {
    const path: number[] = [];
    const res: number[][] = [];
    let sum = 0;
    
    candidates.sort((a, b) => a - b);
    
    function backtrack(start: number) : void {
        if (sum === target) {
            res.push([...path]);
            return;
        }
        
        for (let i = start; i < candidates.length; i++) {
            if (sum + candidates[i] > target) {
                return;
            }
            
            path.push(candidates[i]);
            sum += candidates[i];
            backtrack(i);
            sum -= candidates[i];
            path.pop();
        }
    }

    backtrack(0);
    return res;
};